Lesson2a

paulcraig opened the pull request:
I built a Jupyter book that included the Introduction and File Parsing notebooks. It looked good on my computer. I fixed one typo. I think it is in good shape.

paulcraig committed:
added File Parsing to the TOC

paulcraig committed:
corrected one typo

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Absolute and relative paths\n",
```
In the original lessons this was listed as a note, or aside. It simply shows up as a section in this version.

paulcraig replied:
fixed.

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outfile = open(protein_file,\"r\")\n",
```
The preferred method to do this would actually be to do 

    with open(protein_file, "r") as f:
        data = f.readlines()
     
We have been considering switching this in the python scripting lesson for a while. If this syntax is used, you do not have to remember to close the file and Python does it automatically.

paulcraig replied:
Done

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## An alternative way to open a file.\n",
```
This alternative way is actually the preferred way by most people. Would you be alright with using this syntax in the lesson?

paulcraig replied:
Done

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "outputs": [],
+   "source": [
+    "for line in data:\n",
+    "    if 'HETNAM' in line:\n",
```
does this happen several times? (as in, lines with 'HETNAM' appears several times?). In that case, many lines will be printed, but only the last one will be stored. Is there any line which will occur only once?

paulcraig replied:
In 4eyr.pdb, there is only one HETNAM line, representing the single non-water heterogroup in the structure, so this works with 4eyr.pdb. In a file such as 2w9g.pdb, there are three lines that start with HETNAM, representing three non-water heterogroups in the structure, so the three lines would print, but only the last one would be stored as HETNAM_line. I can’t think of a line that is guaranteed to occur only once in a PDB file. This is actually a flaw in the PDB file format and one of the reasons that the PDB has switched to cif format for their files. I think the example here serves the purpose of the notebook. If someone wants to dig deeper, they will encounter the problem you mentioned. Should I include a note on how to change the loop if there is more than one HETNAM line?

janash replied:
Since there can be more than one `HETNAM`, do you see any value in keeping a list of them? Perhaps this is a case where you would want to create an empty list outside of the loop, and append the `HETATM` lines. This could even be an exercise - "more than one `HETNAM` line can occur in a pdb file, can you think of a way to keep all of the lines using syntax we have already learned?" (they have learned to create an empty list and append to it in the first lesson)

paulcraig replied:
DONE. The output is not elegant, but it contains all of the heterogen names and abbreviations as a list.

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "metadata": {},
+   "source": [
+    "<div class=\"exercise-this-is-a-title exercise\">\n",
+    "<p class=\"exercise-title\">Exercise on file parsing</p>\n",
```
This exercise strikes me as being a jump from the material, and does not seem like something you would ask the students to do during the workshop (it's a lot of typing). Is this seen as a "homework" assignment/do you anticipate using it during a live workshop?

paulcraig replied:
I modified and shortened the exercise to pull different strings out of 4eyr.pdb instead of adding a new file, 7tim.pdb. I think it could work in a workshop now.

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
````diff
@@ -0,0 +1,602 @@
+    "<p>This is one possible solution for the 7tim.pdb parsing exercise.</p>\n",
+    "    \n",
+    "```python\n",
+    "with open(protein_file,\"r\") as outfile2:\n",
````
The technique we listed as an "alternate" technique for opening files is used here. Either it needs to use `open` and `close`, or the earlier portion of the lesson needs to use the same syntax as this solution.

paulcraig replied:
I changed the earlier portion of the lesson to include the with context manager.

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+    "\n",
+    "When you use a for loop, it is easy to have python keep up with the line numbers using the `enumerate` command.  The general syntax is\n",
+    "\n",
+    "`for linenum, line in enumerate(list_name):\n",
```
Use three ticks for code block. Both of these are showing up on the same line now.

paulcraig replied:
Fixed

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
````diff
@@ -0,0 +1,602 @@
+    "    \n",
+    "```python\n",
+    "with open(protein_file,\"r\") as outfile2:\n",
+    "    data2 = outfile2.readlines()\n",
````
I would also advise against using `data2` variable names, though it seems convenient. Do we really need to switch PDB files halfway through the lesson? It seems more consistent to me to work with only one, and to move on to a second/third/fourth during the multiple file parsing lesson. 

When a new file is read within an exercise solution, you may run into a few problems. For example, in a real workshop you might not have time for this exercise since it's complex and a little extensive. I think it would be a better idea to finish the lesson with the original pdb.

paulcraig replied:
I switched the exercise to just search for two different strings in 4eyr.pdb, so that I was able to remove 7tim.pdb from this notebook and stayed with data instead of data2.

janash commented on biochemist-python/chapters/File_Parsing.ipynb:
```diff
@@ -0,0 +1,602 @@
+   "metadata": {},
+   "source": [
+    "## A final note about regular expressions\n",
+    "Sometimes you will need to match something more complex than just a particular word or phrase in your output file.  Sometimes you will need to match a particular word, but only if it is found at the beginning of a line.  Or perhaps you will need to match a particular pattern of data, like a capital letter followed by a number, but you won't know the exact letter and number you are looking fo ...
```
Missing link on the last 'here'.

paulcraig replied:
Fixed

janash reviewed (changes requested):
Overall looks good. I've made some comments at specific places. 

There are three main discussion points or requests for changes:

1.  Would you like to introduce opening files using a context manager (that is to say, using the word `with`) from the beginning? This is actually the preferred way to open files, and prevents you from having to remember to close the file. Although it is not explicitly introduced in the lesson, you do use it in an exercise solution.
2. I think it would be better only work with one file in the file parsing lesson.
3. The files are not in this pull request, so I was not actually able to run the code. Can you commit them to this branch so I can run the code?

Please see all comments in the review. Thanks!

paulcraig committed:
adding the two PDB files for the workbook

paulcraig committed:
with suggested corrections

paulcraig commented:
Hi Jessica,

I made the corrections you suggested, detailed in yellow below in response to each correction. I need to learn to do three things:


  1.  Delete a file from a local and Github repository. I need to remove 7tim.pdb.
  2.  Issue a new pull request. I pushed the corrected File_parsing.ipynb to Github, but it did not generate a pull request this time.
  3.  I just went on Github and entered my responses there. Can you see the changed file or do I need to issue a new pull request?

Paul

Paul A. Craig
Professor of Biochemistry
School of Chemistry & Materials Science
Rochester Institute of Technology
1 Lomb Memorial Drive
Building 08, Room A114
Rochester, NY 14623
Telephone: +1.585.475.6145


From: "Jessica A. Nash" <notifications@github.com>
Reply-To: MolSSI-Education/python-scripting-biochemistry <reply@reply.github.com>
Date: Sunday, March 7, 2021 at 7:38 PM
To: MolSSI-Education/python-scripting-biochemistry <python-scripting-biochemistry@noreply.github.com>
Cc: Paul A Craig <pac8612@rit.edu>, Author <author@noreply.github.com>
Subject: Re: [MolSSI-Education/python-scripting-biochemistry] Lesson2a (#3)


@janash requested changes on this pull request.

Overall looks good. I've made some comments at specific places.

There are three main discussion points or requests for changes:

  1.  Would you like to introduce opening files using a context manager (that is to say, using the word with) from the beginning? This is actually the preferred way to open files, and prevents you from having to remember to close the file. Although it is not explicitly introduced in the lesson, you do use it in an exercise solution.

Paul: That makes sense to me. First I introduced the simple method that requires the closefile command, then I added the preferred version.

  2.  I think it would be better only work with one file in the file parsing lesson.

Paul: I switched the exercise to pull some additional strings from 4eyr.pdb and removed 7tim.pdb from the notebook.

  3.  The files are not in this pull request, so I was not actually able to run the code. Can you commit them to this branch so I can run the code?

Paul: I added data/PDB_files/4eyr.pdb. I had also added 7tim.pdb, but I’m not using that anymore. How can I remove it from my local and Github repositories?

Please see all comments in the review. Thanks!

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589111063>:

> +    "\n",

+    "data/PDB_files/4eyr.pdb"

+   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "Here, we have specified that our filepath contains the 'data' and 'PDB_files' directory, and the `os.path` module has made this into a filepath that is usable by our system. If you are on Windows, you will instead see that a forward slash is used."

+   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "## Absolute and relative paths\n",

In the original lessons this was listed as a note, or aside. It simply shows up as a section in this version. FIXED

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589111488>:

> +  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "## Reading a file\n",

+    "\n",

+    "In Python, there are many ways to read in information from a text file.  The best method to use depends on the type of data and the type of analysis you are performing.  If you have a file with lots of different types of information, text and numbers, with different types of formatting, the most generic way to read in information is the `readlines()` function.  Before you can read in a file, you have to open the file using the file path we defined above. This will create a file object, or filehandle.  The file we will be analyzing in this example is a [PDB web site](https://www.rcsb.org/structure/4EYR) file for an HIV protease complex with the inhibitor Ritonavir."

+   ]

+  },

+  {

+   "cell_type": "code",

+   "execution_count": null,

+   "metadata": {},

+   "outputs": [],

+   "source": [

+    "outfile = open(protein_file,\"r\")\n",

The preferred method to do this would actually be to do

with open(protein_file, "r") as f:

    data = f.readlines()

We have been considering switching this in the python scripting lesson for a while. If this syntax is used, you do not have to remember to close the file and Python does it automatically.  DONE

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589111791>:

> +   ]

+  },

+  {

+   "cell_type": "code",

+   "execution_count": null,

+   "metadata": {},

+   "outputs": [],

+   "source": [

+    "outfile.close()"

+   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "## An alternative way to open a file.\n",

This alternative way is actually the preferred way by most people. Would you be alright with using this syntax in the lesson?  DONE

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589112312>:

> +   "source": [

+    "This will print exactly what is in the file.\n",

+    "\n",

+    "If you look through the output, you will find a line that starts with HETNAM. This line contains information about the ligand or small molecule that is bound to HIV protease in this structure, providing both the abbreviation (RIT) and the full name of the drug (RITONAVIR).  We want to search through this file and find that line, and print only that line. We can do this using an `if` statement.\n",

+    "\n",

+    "Returning to our file example,"

+   ]

+  },

+  {

+   "cell_type": "code",

+   "execution_count": null,

+   "metadata": {},

+   "outputs": [],

+   "source": [

+    "for line in data:\n",

+    "    if 'HETNAM' in line:\n",

does this happen several times? (as in, lines with 'HETNAM' appears several times?). In that case, many lines will be printed, but only the last one will be stored. Is there any line which will occur only once?

Paul: In 4eyr.pdb, there is only one HETNAM line, representing the single non-water heterogroup in the structure, so this works with 4eyr.pdb. In a file such as 2w9g.pdb, there are three lines that start with HETNAM, representing three non-water heterogroups in the structure, so the three lines would print, but only the last one would be stored as HETNAM_line. I can’t think of a line that is guaranteed to occur only once in a PDB file. This is actually a flaw in the PDB file format and one of the reasons that the PDB has switched to cif format for their files. I think the example here serves the purpose of the notebook. If someone wants to dig deeper, they will encounter the problem you mentioned. Should I include a note on how to change the loop if there is more than one HETNAM line?

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589112632>:

> +  },

+  {

+   "cell_type": "code",

+   "execution_count": null,

+   "metadata": {},

+   "outputs": [],

+   "source": [

+    "atoms = float(words[1])"

+   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "<div class=\"exercise-this-is-a-title exercise\">\n",

+    "<p class=\"exercise-title\">Exercise on file parsing</p>\n",

This exercise strikes me as being a jump from the material, and does not seem like something you would ask the students to do during the workshop (it's a lot of typing). Is this seen as a "homework" assignment/do you anticipate using it during a live workshop?

Paul: I modified and shortened the exercise to pull different strings out of 4eyr.pdb instead of adding a new file, 7tim.pdb.

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589112769>:

> +   "source": [

+    "<div class=\"exercise-this-is-a-title exercise\">\n",

+    "<p class=\"exercise-title\">Exercise on file parsing</p>\n",

+    "<p>Use the provided 7tim.pdb file, which includes structural information and data about a complex of of a small molecule bound to the enzyme, triose phosphate isomerase. Find the resolution of the structure, teh total number of protein atoms, the total number of heterogroup atoms, and the identify the small molecule (hint: HETNAM) by abbreviation and full name. Remember - you need to use os.join to be able to read the file. Your output should look like this:</p>\n",

+    "\n",

+    "<p>RESOLUTION : 1.90 ANGSTROMS  </p>  \n",

+    "<p>PROTEIN ATOMS : 3766  </p>\n",

+    "<p>HETEROGEN ATOMS : 20  </p>\n",

+    "<p>HETNAM Abbreviation: PGH Full name: PHOSPHOGLYCOLOHYDROXAMIC </p>\n",

+    "    \n",

+    "```{admonition} Solution\n",

+    ":class: dropdown\n",

+    "<p>This is one possible solution for the 7tim.pdb parsing exercise.</p>\n",

+    "    \n",

+    "```python\n",

+    "with open(protein_file,\"r\") as outfile2:\n",

The technique we listed as an "alternate" technique for opening files is used here. Either it needs to use open and close, or the earlier portion of the lesson needs to use the same syntax as this solution.

Paul: I changed the earlier portion of the lesson to include the with context manager.

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589112916>:

> +   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "## Searching for a particular line number in your file\n",

+    "There is a lot of other information in the PDB file we might be interested in.  For example, we might want to pull out the sequence of the protein.  If we look through the file in a text editor, we notice that the sequence is given in a series of lines that begin with `SEQRES` \n",

+    "\n",

+    "SEQRES   1 A  247  ALA ARG THR PHE PHE VAL GLY GLY ASN PHE LYS LEU ASN  \n",

+    "\n",

+    "followed by a seriew of lines that contain the full sequence using three letter abbreviations for the amino acids.  In this case, we don't want to pull something out of this line, as we did in our previous example, but we want to know which line of the file this is so that we can then pull the sequence from the next few lines.  \n",

+    "\n",

+    "When you use a for loop, it is easy to have python keep up with the line numbers using the `enumerate` command.  The general syntax is\n",

+    "\n",

+    "`for linenum, line in enumerate(list_name):\n",

Use three ticks for code block. Both of these are showing up on the same line now.  DONE

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589113410>:

> +    "<div class=\"exercise-this-is-a-title exercise\">\n",

+    "<p class=\"exercise-title\">Exercise on file parsing</p>\n",

+    "<p>Use the provided 7tim.pdb file, which includes structural information and data about a complex of of a small molecule bound to the enzyme, triose phosphate isomerase. Find the resolution of the structure, teh total number of protein atoms, the total number of heterogroup atoms, and the identify the small molecule (hint: HETNAM) by abbreviation and full name. Remember - you need to use os.join to be able to read the file. Your output should look like this:</p>\n",

+    "\n",

+    "<p>RESOLUTION : 1.90 ANGSTROMS  </p>  \n",

+    "<p>PROTEIN ATOMS : 3766  </p>\n",

+    "<p>HETEROGEN ATOMS : 20  </p>\n",

+    "<p>HETNAM Abbreviation: PGH Full name: PHOSPHOGLYCOLOHYDROXAMIC </p>\n",

+    "    \n",

+    "```{admonition} Solution\n",

+    ":class: dropdown\n",

+    "<p>This is one possible solution for the 7tim.pdb parsing exercise.</p>\n",

+    "    \n",

+    "```python\n",

+    "with open(protein_file,\"r\") as outfile2:\n",

+    "    data2 = outfile2.readlines()\n",

I would also advise against using data2 variable names, though it seems convenient. Do we really need to switch PDB files halfway through the lesson? It seems more consistent to me to work with only one, and to move on to a second/third/fourth during the multiple file parsing lesson.

When a new file is read within an exercise solution, you may run into a few problems. For example, in a real workshop you might not have time for this exercise since it's complex and a little extensive. I think it would be a better idea to finish the lesson with the original pdb.

Paul: I switched the exercise to just search for two different strings in 4eyr.pdb, so that I was able to remove 7tim.pdb from this notebook and stayed with data instead of data2.

________________________________

In biochemist-python/chapters/File_Parsing.ipynb<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#discussion_r589113561>:

> +    "SEQRES   4 A   99  GLY ARG TRP LYS PRO LYS LEU ILE GLY GLY ILE GLY GLY          \n",

+    "\n",

+    "SEQRES   5 A   99  PHE VAL LYS VAL ARG GLN TYR ASP GLN VAL PRO ILE GLU          \n",

+    "\n",

+    "SEQRES   6 A   99  ILE CYS GLY HIS LYS VAL ILE GLY THR VAL LEU VAL GLY          \n",

+    "```\n",

+    "</div>\n",

+    "    "

+   ]

+  },

+  {

+   "cell_type": "markdown",

+   "metadata": {},

+   "source": [

+    "## A final note about regular expressions\n",

+    "Sometimes you will need to match something more complex than just a particular word or phrase in your output file.  Sometimes you will need to match a particular word, but only if it is found at the beginning of a line.  Or perhaps you will need to match a particular pattern of data, like a capital letter followed by a number, but you won't know the exact letter and number you are looking for.  These types of matching situations are handled with something called *regular expressions* which is accessed through the python module `re`.  While using regular expressions is outside the scope of this tutorial, they are very useful and you might want to learn more about them in the future.  A tutorial can be found at [Automate the Boring Stuff with Python book](https://automatetheboringstuff.com/2e/chapter7/). A great test site for regex is [here]()."

Missing link on the last 'here'.  FIXED

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/MolSSI-Education/python-scripting-biochemistry/pull/3#pullrequestreview-605901944>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAXUQDSKT55KKHZIACXRTFLTCQMBHANCNFSM4YRWMRXQ>.

janash commented:
1.  Delete a file from a local and Github repository. I need to remove 7tim.pdb.

If you remove a file you should be able to add and commit that removal. You can also use the commang `git rm` (https://git-scm.com/docs/git-rm)  

  2.  Issue a new pull request. I pushed the corrected File_parsing.ipynb to Github, but it did not generate a pull request this time.
  
  When you push to a branch which has an open pull request the open pull request is automatically updated. You should not see a new one, the old one should be updated. You can check this on the "commits" tab of the pull request.
  
  3.  I just went on Github and entered my responses there. Can you see the changed file or do I need to issue a new pull request?

I can see that the files have changed. @paulcraig

paulcraig committed:
added a new check your understanding

paulcraig committed:
added a file with multiple HETNAM entries

janash commented:
These changes all look good. I like the exercise where we use appending to a list for the `HETNAM` entries. One comment is that all of the code that does manipulation of the file contents does not have to be under `with`. 

For example, in the solution

```
# Use os to assign the file path to a variable
import os
protein_file = os.path.join('data', 'PDB_files','1ddo.pdb')

# Create a list to hold the heterogen information
HETNAM_list = []

# Use the with context manager to open the file, then a for loop to populate HETNAM_list[]
with open(protein_file,"r") as outfile:
    data = outfile.readlines()
    for line in data:
        if 'HETNAM' in line:
            words = line.split()
            HETNAM_list.append(words[1:3])
print(HETNAM_list) 
```

We use a `for` loop to go through the file and it is under the `with` statement. Alternatively, you could also do:

```
# Use os to assign the file path to a variable
import os
protein_file = os.path.join('data', 'PDB_files','1ddo.pdb')

# Create a list to hold the heterogen information
HETNAM_list = []

# Use the with context manager to open the file, then a for loop to populate HETNAM_list[]
with open(protein_file,"r") as outfile:
    data = outfile.readlines()

for line in data:
    if 'HETNAM' in line:
        words = line.split()
        HETNAM_list.append(words[1:3])

print(HETNAM_list) 
```

This separates out the file reading (under the `with` statement) from working with what is in the file. I prefer this as well because then you do not have as much indentation. The major difference here is that the `for` in unindented.

I'm also wondering if you meant to append `words[1:3]`, or if you meant to append `words[1]` as you previously had them grab. The current code would result in a nested list (list of lists). 

I promise that we're really close to merging this @paulcraig :)

paulcraig commented:
> :)

I'm also wondering if you meant to append words[1:3], or if you meant to append words[1] as you previously had them grab. The current code would result in a nested list (list of lists).

Yes. I want them to see the pairs of HETNAM abbreviations and longer names when the structure contains more than one heterogen. 

I'm curious - why couldn't I reply to the message you sent this time. I had to create a new message.

janash committed:
Merge branch 'main' into lesson2a
